cpp/src/io/json/json_tree.cu

```diff
@@ -439,210 +457,227 @@ rmm::device_uvector<size_type> hash_node_type_with_field_name(device_span<Symbol
+                      col_id.end(),
+                      col_id.begin());
+
+  rmm::device_uvector<size_type> parent_col_id(num_nodes, stream, mr);
+  thrust::transform(rmm::exec_policy(stream),
+                    d_tree.parent_node_ids.begin(),
+                    d_tree.parent_node_ids.end(),
+                    parent_col_id.begin(),
+                    [col_id = col_id.begin()] __device__(auto node_id) {
+                      return node_id >= 0 ? col_id[node_id] : parent_node_sentinel;
+                    });
 
   return {std::move(col_id), std::move(parent_col_id)};
 }
 
 /**
  * @brief Computes row indices of each node in the hierarchy.
- * 5. Generate row_offset.
- *   a. stable_sort by parent_col_id.
- *   b. scan_by_key {parent_col_id} (required only on nodes who's parent is list)
- *   c. propagate to non-list leaves from parent list node by recursion
+ * 2. Generate row_offset.
+ *   a. filter only list childs
+ *   b. stable_sort by parent_col_id.
+ *   c. scan_by_key {parent_col_id} (done only on nodes who's parent is list)
+ *   d. propagate to non-list leaves from parent list node by recursion
  *
  * pre-condition:
- *  scatter_indices is a sequence, representing node_id.
  *  d_tree.node_categories, d_tree.parent_node_ids, parent_col_id are in order of node_id.
  * post-condition: row_offsets is in order of node_id.
- *  parent_col_id and scatter_indices are sorted by parent_col_id. (unused after this function)
- * @param scatter_indices node_id
+ *  parent_col_id is moved and reused inside this function.
  * @param parent_col_id parent node's column id
  * @param d_tree Tree representation of the JSON string
  * @param stream CUDA stream used for device memory operations and kernel launches.
  * @param mr Device memory resource used to allocate the returned column's device memory.
  * @return row_offsets
  */
-rmm::device_uvector<size_type> compute_row_offsets(device_span<size_type> scatter_indices,
-                                                   rmm::device_uvector<NodeIndexT>&& parent_col_id,
-                                                   tree_meta_t& d_tree,
+rmm::device_uvector<size_type> compute_row_offsets(rmm::device_uvector<NodeIndexT>&& parent_col_id,
+                                                   tree_meta_t const& d_tree,
                                                    rmm::cuda_stream_view stream,
                                                    rmm::mr::device_memory_resource* mr)
 {
   CUDF_FUNC_RANGE();
   auto const num_nodes = d_tree.node_categories.size();
-  // TODO generate scatter_indices sequences here itself
-  thrust::stable_sort_by_key(
-    rmm::exec_policy(stream), parent_col_id.begin(), parent_col_id.end(), scatter_indices.begin());
+
+  rmm::device_uvector<size_type> scatter_indices(num_nodes, stream);
+  thrust::sequence(rmm::exec_policy(stream), scatter_indices.begin(), scatter_indices.end());
+
+  // filter only list childs. (nodes who's parent is a list/root)
```

Comment:
`filter` is ambiguous, it could mean "filter (into)" or "filter (from)", `remove` would probably be more appropriate

Reply:
updated.